core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Path
+    //////////////////////////////////
+
+    public XContentBuilder value(Path value) throws IOException {
         if (value == null) {
             return nullValue();
         }
-        generator.writeBinary(value, offset, length);
-        return this;
+        return value(value.toString());
     }
 
-    /**
-     * Writes the binary content of the given BytesRef
-     * Use {@link org.elasticsearch.common.xcontent.XContentParser#binaryValue()} to read the value back
-     */
-    public XContentBuilder value(BytesRef value) throws IOException {
-        if (value == null) {
+    ////////////////////////////////////////////////////////////////////////////
+    // Objects
+    //
+    // These methods are used when the type of value is unknown. It tries to fallback
+    // on typed methods and use Object.toString() as a last resort. Always prefer using
+    // typed methods over this.
+    //////////////////////////////////
+
+    public XContentBuilder field(String name, Object value) throws IOException {
+        return field(name).value(value);
+    }
+
+    public XContentBuilder array(String name, Object... values) throws IOException {
+        return field(name).values(values);
+    }
+
+    XContentBuilder values(Object[] values) throws IOException {
+        if (values == null) {
             return nullValue();
         }
-        generator.writeBinary(value.bytes, value.offset, value.length);
+        startArray();
+        for (Object o : values) {
+            value(o);
+        }
+        endArray();
         return this;
     }
 
-    /**
-     * Writes the binary content of the given BytesReference
-     * Use {@link org.elasticsearch.common.xcontent.XContentParser#binaryValue()} to read the value back
-     */
-    public XContentBuilder value(BytesReference value) throws IOException {
-        if (value == null) {
-            return nullValue();
-        }
-        BytesRef ref = value.toBytesRef();
-        generator.writeBinary(ref.bytes, ref.offset, ref.length);
+    public XContentBuilder value(Object value) throws IOException {
+        unknownValue(value);
         return this;
     }
 
-    public XContentBuilder value(Text value) throws IOException {
+    private void unknownValue(Object value) throws IOException {
         if (value == null) {
-            return nullValue();
-        } else if (value.hasString()) {
-            generator.writeString(value.string());
+            nullValue();
+            return;
+        }
+        Writer writer = WRITERS.get(value.getClass());
+        if (writer != null) {
+            writer.write(this, value);
+        } else if (value instanceof Path) {
+            //Path implements Iterable<Path> and causes endless recursion and a StackOverFlow if treated as an Iterable here
+            value((Path) value);
```

Comment:
Yeah. I'm not sure either. I like that concrete classes are nice and quick. Lets drop this point for this PR at least.